cli/open.go

```diff
@@ -0,0 +1,268 @@
+
+			// We need a started workspace to figure out e.g. expanded directory.
+			// Pehraps the vscode-coder extension could handle this by accepting
+			// default_directory=true, then probing the agent. Then we wouldn't
+			// need to wait for the agent to start.
+			workspaceQuery := inv.Args[0]
+			autostart := true
+			workspace, workspaceAgent, err := getWorkspaceAndAgent(ctx, inv, client, autostart, codersdk.Me, workspaceQuery)
+			if err != nil {
+				return xerrors.Errorf("get workspace and agent: %w", err)
+			}
+
+			workspaceName := workspace.Name + "." + workspaceAgent.Name
+			insideThisWorkspace := insideAWorkspace && inWorkspaceName == workspaceName
+
+			if !insideThisWorkspace {
+				// We could optionally add a flag to skip wait, like with SSH.
+				wait := false
+				for _, script := range workspaceAgent.Scripts {
+					if script.StartBlocksLogin {
+						wait = true
+						break
+					}
+				}
+				err = cliui.Agent(ctx, inv.Stderr, workspaceAgent.ID, cliui.AgentOptions{
+					Fetch:     client.WorkspaceAgent,
+					FetchLogs: client.WorkspaceAgentLogsAfter,
+					Wait:      wait,
+				})
+				if err != nil {
+					if xerrors.Is(err, context.Canceled) {
+						return cliui.Canceled
+					}
+					return xerrors.Errorf("agent: %w", err)
+				}
+
+				// If the ExpandedDirectory was initially missing, it could mean
+				// that the agent hadn't reported it in yet. Retry once.
+				if workspaceAgent.ExpandedDirectory == "" {
+					autostart = false // Don't retry autostart.
+					workspace, workspaceAgent, err = getWorkspaceAndAgent(ctx, inv, client, autostart, codersdk.Me, workspaceName)
+					if err != nil {
+						return xerrors.Errorf("get workspace and agent retry: %w", err)
+					}
+				}
+			}
+
+			directory := workspaceAgent.ExpandedDirectory // Empty unless agent directory is set.
+			if len(inv.Args) > 1 {
+				d := inv.Args[1]
+
+				switch {
+				case insideThisWorkspace:
+					// TODO(mafredri): Return error if directory doesn't exist?
+					directory, err = filepath.Abs(d)
+					if err != nil {
+						return xerrors.Errorf("expand directory: %w", err)
+					}
+
+				case d == "~" || strings.HasPrefix(d, "~/"):
+					return xerrors.Errorf("path %q requires expansion and is not supported, use an absolute path instead", d)
+
+				case workspaceAgent.OperatingSystem == "windows":
+					switch {
+					case directory != "" && !isWindowsAbsPath(d):
+						directory = windowsJoinPath(directory, d)
+					case isWindowsAbsPath(d):
+						directory = d
+					default:
+						return xerrors.Errorf("path %q not supported, use an absolute path instead", d)
+					}
+
+				// Note that we use `path` instead of `filepath` since we want Unix behavior.
+				case directory != "" && !path.IsAbs(d):
+					directory = path.Join(directory, d)
+				case path.IsAbs(d):
+					directory = d
+				default:
+					return xerrors.Errorf("path %q not supported, use an absolute path instead", d)
+				}
+			}
```

Comment:
I think this could be extracted to a helper function and tested individually.

Reply:
It's a good idea, I was planning on doing that once we add another command, like jetbrains gateway or something. Much of this should be shareable between the two.

Reply:
Maybe we can leave this till it's needed then!